### PR TITLE
[GO] Examples cleanup

### LINKED ISSUE #1357: [Go] Example revamp
## 📖 Description
We need to cleanup the examples as well as make them more useful for adopters.
- one-of-everything needs its authz logic fixed
- one-of-everything needs struct example
- hello_world basic catena device setup example
- remove all the other examples

---

## 🎯 Goal / Outcome
An adopter should be able to get set up easily with the hello_world example, as well as have any possible functionality shown in one_of_everything.

---

## ✅ Definition of Done (DoD)
- [ ] Feature is implemented and works as expected
- [ ] Edge cases are handled
- [ ] Tests are added or updated (if applicable)
- [ ] Documentation is updated (if needed)
- [ ] Code is reviewed and approved

---

## 🛠️ Implementation Notes (Optional)

---

## 🔗 Related Issues / Links

---
## 🚧 Risks / Open Questions (Optional)

---

### PR DESCRIPTION
**Pull request by ctwarog-ross**
## 📖 Description
Closes Issue #1357 

Removed all examples other than one-of-everything
one-of-everything has struct implementation
one-of-everything authz fixed
one-of-everything now has all the implemented endpoint edge cases

hello_world example
hello_world contains a string with get/setvalue endpoint exposed
runs on rest and grpc

---

## ✅ Definition of Done (DoD)

---

## 🛠️ Implementation Notes

---

## 🔗 Related Issues / Links

---
## 📝 Additional Context

---

## 🚧 Risks / Open Questions

---
## ✨ Updates
<!-- (Optional) List summarizing changes made when editing the pull request>

### CODE REVIEW

File: sdks/go/examples/hello_world/main.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+func buildDevice() map[string]any {
+	return map[string]any{
+		"slot":              uint32(slot),
+		"detail_level":      catena.DetailLevelFull,
+		"multi_set_enabled": true,
+		"subscriptions":     true,
+		"access_scopes":     []string{"st2138:mon", "st2138:op", "st2138:cfg", "st2138:adm"},
+		"default_scope":     "st2138:op",
+		"params": map[string]any{
```
Comment (abrown-RV) [inline sdks/go/examples/hello_world/main.go:84]: should get the mandatory params... but this will be taken care of by the helpers, so meh?

File: sdks/go/examples/hello_world/main.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+		logger.Info("GetDevice", "slot", slot)
+		device, err := catena.ToDevice(buildDevice())
+		if err != nil {
+			logger.Error("Failed to convert device", "slot", slot, "error", err)
+			return catena.ReplyError[catena.Device](catena.INTERNAL, "failed to convert device")
+		}
+		return catena.Reply(device)
+	})
+
+	srv.RegisterGetValueHandler(slot, func(slot uint16, fqoid string, ctx catena.HandlerContext) (catena.Value, catena.StatusResult) {
```
Comment (abrown-RV) [inline sdks/go/examples/hello_world/main.go:120]: should check for `op` scope, since that's the default, above and `hello_world` does not specify a different scope. Or just hardcode AuthzEnabled to false since this is a basic example

File: sdks/go/examples/hello_world/main.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+			return catena.StatusWithCode(catena.NOT_FOUND, "parameter not found: "+fqoid)
+		}
+
+		stringValue, ok := value.(string)
+		if !ok {
+			return catena.StatusWithCode(catena.INVALID_ARGUMENT, "hello_world requires a string value")
+		}
+
+		helloWorld.Set(stringValue)
+		srv.BroadcastUpdate(slot, helloWorldOID, stringValue, catena.ScopeMon)
```
Comment (abrown-RV) [inline sdks/go/examples/hello_world/main.go]: should check for `ScopeOp` write at the top of `SetValue`, and broadcast on `ScopeOp`

File: sdks/go/examples/oneOfEverything/main.go [hunk 4]
```diff
@@ -157,6 +158,252 @@ func stringParamValue(p *sync.Map, key string, fallback string) map[string]any {
...
+		if !ok {
+			return nil, false
+		}
+		params = childParamMap(paramInfo)
+	}
+
+	return paramInfo, true
+}
+
+func paramInfosForRequest(params map[string]any, oid string, recursive bool) ([]catena.ParamInfo, catena.StatusResult) {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/main.go]: all this extra plumbing for the param infos is maybe an indicator that we could be doing them nicer, but something to think about.

File: sdks/go/examples/oneOfEverything/main.go
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/main.go:1]: We should get better commenting on these examples for them to be good for people to look at

**Review by abrown-RV — COMMENTED** [review]
Do you want to spit up oneofeverthing's main.go into different files for the different endpoints here? or is that another PR?

File: sdks/go/examples/hello_world/main.go (hunk 2 above)
Comment (ctwarog-ross) [reply sdks/go/examples/hello_world/main.go:120]: Authzenabled to false is a good idea

File: sdks/go/examples/hello_world/main.go (hunk 3 above)
Comment (ctwarog-ross) [reply sdks/go/examples/hello_world/main.go]: We'll leave authz out of the hello_world example

**Comment by ctwarog-ross** [comment]
bugbot run

**Comment by ctwarog-ross** [comment]
bugbot run

File: sdks/go/examples/hello_world/main.go (hunk 3 above)
Comment (abrown-RV) [reply sdks/go/examples/hello_world/main.go]: makes sense, should just pass "" for the scope then? ... we need a good convention for what to pass when authz is disabled. It doesn't matter but like empty string at least saves some bytes? XD

File: sdks/go/pkg/catena/param_info.go [hunk 5]
```diff
@@ -89,6 +91,157 @@ func ToParamInfo(m map[string]any) (ParamInfo, error) {
...
+			name.With(lang, s)
+		}
+	}
+	if len(name) == 0 {
+		return nil
+	}
+	return name
+}
+
+func paramArrayLength(paramInfo map[string]any) uint32 {
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/param_info.go:134]: where is this coming from, array_length and array_len aren't valid fields in a device definition?

File: sdks/go/examples/oneOfEverything/access_handler.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+func registerAccessHandler(srv catena.Server) {
+	// AccessHandler is the coarse authorization gate for SDK endpoints. It runs
+	// after the transport has built HandlerContext, so adopters can check scopes,
+	// identity, or endpoint type before any endpoint-specific handler runs. Use
+	// this for cross-cutting policy; use individual handlers for per-param rules.
+	srv.RegisterAccessHandler(func(endpointType catena.EndpointType, ctx catena.HandlerContext) bool {
+		logger.Info("Access request", "endpointType", endpointType)
+
+		// Shows how to restrict getPopulatedSlots to only op and adm read scopes.
+		if endpointType == catena.EndpointGetSlots {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/access_handler.go]: I don't think we should do this. Struggling to come up with a good demo to put here, maybe its enough to just have it return true and have all the comments?

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+		} else if strings.HasPrefix(key, "sample_float") && key == "sample_float" {
+			v, ok = state.sampleFloat, true
+		}
+		state.mu.RUnlock()
+
+		return replyValue(fqoid, v, ok)
+	})
+
+	// Slot 0: writes use the same manual switch strategy as reads. Writes are
+	// restricted to callers with configuration-scope write access.
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: might be good to have examples of params with individual scopes instead of blanket accross a whole slot.

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+
+		var v any
+		var ok bool
+		if strings.HasPrefix(key, "product") {
+			v, ok = state.slotTwoProductValue(key)
+			return replyValue(fqoid, v, ok)
+		}
+
+		state.mu.RLock()
+		if strings.HasPrefix(key, "volume") && key == "volume" {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: isn't the HasPrefix redundant?

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+		} else if strings.HasPrefix(key, "struct_example") {
+			switch key {
+			case "struct_example":
+				v, ok = cloneStringAnyMap(state.structExample), true
+			case "struct_example/number":
+				v, ok = state.structExample["number"], true
+			case "struct_example/text":
+				v, ok = state.structExample["text"], true
+			}
+		} else if strings.HasPrefix(key, "sample_struct_variant_array") && key == "sample_struct_variant_array" {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: all of the rest of these need to be broken down by sub-params like `struct_example`

friendly reminder that array has /index

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+			v, ok = cloneStructVariantArray(state.sampleStructVariantArray), true
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: why clone? In catena.ToValue() it will be cloned there. Right? putting it into a proto clones it? maybe not... ok looking at the proto code right, its only the scalars that are copied by value, and therefore "cloned".

Maybe it would be helpful to make tovalue clone? since that would be a pretty safe assumption that you don't want it to follow the changing reference? or have it be an option or add a ToValueClone or something?

Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: why clone? In catena.ToValue() it will be cloned there. Right? putting it into a proto clones it? maybe not... ok putting it into the proto doesn't clone it necessarily, but for the most part ToProto will end up cloning everything by how it walks through the heirarchy. I think only for the arrays, it doesn't actually clone.

Maybe it would be helpful to make tovalue consistantly clone? since that would be a pretty safe assumption that you don't want it to follow the changing reference?

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+			v, ok = state.sampleFloat, true
+		}
+		state.mu.RUnlock()
+
+		return replyValue(fqoid, v, ok)
+	})
+
+	// Slot 0: writes use the same manual switch strategy as reads. Writes are
+	// restricted to callers with configuration-scope write access.
+	srv.RegisterSetValueHandler(0, func(value any, slot uint16, fqoid string, ctx catena.HandlerContext) catena.StatusResult {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: the setvalues should check for `product` and reject `PermissionDenied` (its readonly)

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+		if value == nil {
+			logger.Error("SetValue nil value received", "slot", slot, "fqoid", fqoid)
+			return catena.StatusWithCode(catena.StatusCodeInvalidArgument, "nil value received")
+		}
+
+		key := normalizeFqoid(fqoid)
+		switch key {
+		case "running":
+			logger.Warning("SetValue rejected for running param", "slot", slot, "fqoid", fqoid)
+			return catena.StatusWithCode(catena.StatusCodeInvalidArgument, "use start/stop commands to change running state")
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: this should be PermissionDenied, `running` is readonly

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+		case "counter":
+			expectedType := reflect.TypeOf(counter.GetValue())
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: does `if v, ok := value.(int32); ok {` work instead of reflecting? I assume reflecting is slower than casting

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+			}
+			state.sampleIntArray = append([]int32(nil), value.([]int32)...)
+			state.mu.Unlock()
+		default:
+			logger.Error("SetValue param not found", "slot", slot, "fqoid", fqoid)
+			return catena.StatusWithCode(catena.StatusCodeNotFound, "param not found: "+fqoid)
+		}
+
+		logger.Info("Parameter updated", "fqoid", fqoid, "value", value)
+		srv.BroadcastUpdate(slot, fqoid, value, catena.ScopeMon)
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: should be `cfg` scope

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+
+		key := normalizeFqoid(fqoid)
+		switch key {
+		case "resolution", "brightness", "contrast", "saturation":
+			current, ok := state.slotOneParams.Load(key)
+			if !ok {
+				logger.Error("SetValue param not found", "slot", slot, "fqoid", fqoid)
+				return catena.StatusWithCode(catena.StatusCodeNotFound, "param not found: "+fqoid)
+			}
+			if reflect.TypeOf(current) != reflect.TypeOf(value) {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: reflection makes sense here 👍

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+
+		logger.Info("Parameter updated", "fqoid", fqoid, "value", value)
+		srv.BroadcastUpdate(slot, fqoid, value, catena.ScopeMon)
+		return catena.StatusWithCode(catena.StatusCodeOk, "")
+	})
+
+	// Slot 2: writes use prefix-based dispatch and keep product params read-only.
+	// Like the slot 2 read handler, this example does not add a per-handler
+	// scope check so adopters can compare scoped and unscoped endpoint patterns.
+	srv.RegisterSetValueHandler(2, func(value any, slot uint16, fqoid string, ctx catena.HandlerContext) catena.StatusResult {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: I think technically the spec does not allow for unscoped params. Currently the schema doesn't actually enforce it so should check with John, but I think scopes are unescapable, at least when authz is on.

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+			return catena.StatusWithCode(catena.StatusCodeNotFound, "param not found: "+fqoid)
+		}
+		if key != "sample_binary" && reflect.TypeOf(current) != reflect.TypeOf(value) {
+			state.mu.Unlock()
+			logger.Error("SetValue type mismatch", "slot", slot, "fqoid", fqoid,
+				"expected", reflect.TypeOf(current), "got", reflect.TypeOf(value))
+			return catena.StatusWithCode(catena.StatusCodeInvalidArgument, "type mismatch")
+		}
+
+		if strings.HasPrefix(key, "volume") && key == "volume" {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: doubled code?

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+		} else if strings.HasPrefix(key, "struct_example") {
+			switch key {
+			case "struct_example":
+				current, ok = state.structExample, true
+			case "struct_example/number":
+				current, ok = state.structExample["number"], true
+			case "struct_example/text":
+				current, ok = state.structExample["text"], true
+			}
+		} else if strings.HasPrefix(key, "sample_struct_variant_array") && key == "sample_struct_variant_array" {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: these complex types all need to handle having their sub-items set directly

File: sdks/go/examples/oneOfEverything/command_handler.go [hunk 19]
```diff
@@ -0,0 +1,57 @@
+package main
+
+import (
+	"github.com/rossvideo/catena/sdks/go/pkg/catena"
+	"github.com/rossvideo/catena/sdks/go/pkg/logger"
+)
+
+func registerCommandHandler(srv catena.Server, counter *CounterState, broadcastRunning func(), counterScope string) {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/command_handler.go:8]: commands have scope too

File: sdks/go/examples/oneOfEverything/device_handler.go [hunk 20]
```diff
@@ -0,0 +1,364 @@
...
+			WithName(catena.NewPolyglotText("en", "Contrast")).ToMap()
+		saturationParam := catena.NewParamInt32(saturation).
+			WithName(catena.NewPolyglotText("en", "Saturation")).ToMap()
+		return map[string]any{
+			"slot":              uint32(1),
+			"detail_level":      catena.DetailLevelFull,
+			"multi_set_enabled": false,
+			"subscriptions":     true,
+			"access_scopes":     []string{"st2138:mon", "st2138:op", "st2138:cfg", "st2138:adm"},
+			"default_scope":     "st2138:op",
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/device_handler.go]: its mon in value_handlers

File: sdks/go/examples/oneOfEverything/device_handler.go [hunk 21]
```diff
@@ -0,0 +1,364 @@
...
+			WithName(catena.NewPolyglotText("en", "Add 10 to Counter")).ToMap()
+		resetCommand := catena.NewParamEmpty().
+			WithName(catena.NewPolyglotText("en", "Reset Counter")).ToMap()
+		return map[string]any{
+			"slot":              uint32(0),
+			"detail_level":      catena.DetailLevelFull,
+			"multi_set_enabled": true,
+			"subscriptions":     true,
+			"access_scopes":     []string{"st2138:mon", "st2138:op", "st2138:cfg", "st2138:adm"},
+			"default_scope":     "st2138:op",
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/device_handler.go]: value_handlers has it as cfg

File: sdks/go/examples/oneOfEverything/device_handler.go [hunk 22]
```diff
@@ -0,0 +1,364 @@
...
+		}, true
+	case 2:
+		// Slot 2: prefix-dispatched identity/audio params plus FLOAT32, arrays,
+		// BINARY, STRUCT_VARIANT, STRUCT_ARRAY, STRUCT_VARIANT_ARRAY examples.
+		state.mu.RLock()
+		productValue := cloneStringAnyMap(state.slotTwoProduct)
+		volume := state.volume
+		muted := state.muted
+		deviceName := state.deviceName
+		structExample := cloneStringAnyMap(state.structExample)
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/device_handler.go]: same cloning thoughts as in value_handlers

File: sdks/go/examples/oneOfEverything/device_handler.go [hunk 23]
```diff
@@ -0,0 +1,364 @@
...
+		sampleFloat := state.sampleFloat
+		sampleFloatArray := cloneFloat32Slice(state.sampleFloatArray)
+		sampleStringArray := cloneStringSlice(state.sampleStringArray)
+		sampleBinary := cloneBytes(state.sampleBinary)
+		sampleStructVariant := cloneStructVariant(state.sampleStructVariant)
+		sampleStructArray := cloneStructArray(state.sampleStructArray)
+		sampleStructVariantArray := cloneStructVariantArray(state.sampleStructVariantArray)
+		state.mu.RUnlock()
+		productParam := catena.NewParamStruct(productValue).
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/device_handler.go:254]: this is good illustration of actually using the param helpers... and I'm not totally happy... need to think more...

File: sdks/go/examples/oneOfEverything/heartbeat_handler.go [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+	srv.RegisterHeartbeatHandler(1, func(slot uint16) {
+		// Example ticking on the canonical "product/version" param.
+		srv.BroadcastUpdate(1, "product/version", "1.0.0", catena.ScopeMon)
+	})
+	srv.RegisterHeartbeatHandler(2, func(slot uint16) {
+		// Example ticking on a different param to show any can be used for heartbeats.
+		state.mu.RLock()
+		volume := state.volume
+		state.mu.RUnlock()
+		srv.BroadcastUpdate(2, "volume", volume, catena.ScopeMon)
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/heartbeat_handler.go:24]: I think broadcast update should also be made threadsafe with making sure it clones the value its given (it already mostly does cause it also calls ToProto) but then you can just lock and unlock safely around BroadcastUpdate directly

File: sdks/go/examples/oneOfEverything/main.go [hunk 25]
```diff
@@ -911,13 +536,27 @@ func main() {
 
 		logger.Info("Web UI available at:")
 		logger.Info("  http://localhost:8080/")
+		logger.Info("Asset endpoints:")
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/main.go]: these endpoints lists for grpc/rest should either do all the endpoint or none of them

**Review by abrown-RV — CHANGES_REQUESTED** [review]
That sure is a lot of code, looking really good. I'm sure after we've talked all this to death there will be more XD

**Comment by abrown-RV** [comment]
we should remove `normalizeFqoid` from everywhere, since it should be properly doing it now.

File: sdks/go/examples/oneOfEverything/access_handler.go (hunk 6 above)
Comment (ctwarog-ross) [reply sdks/go/examples/oneOfEverything/access_handler.go]: that's a good point, especially since we use getpopulatedslots for other things like generating jwt in postman. Having our flagship demo device having getPopulatedSlots be locked behind specific authz is not ideal

File: sdks/go/examples/oneOfEverything/value_handlers.go (hunk 9 above)
Comment (ctwarog-ross) [reply sdks/go/examples/oneOfEverything/value_handlers.go]: The boilerplate for this got crazy, not sure what to do about this

File: sdks/go/examples/oneOfEverything/value_handlers.go (hunk 13 above)
Comment (ctwarog-ross) [reply sdks/go/examples/oneOfEverything/value_handlers.go]: You are correct, reflection takes significantly longer

File: sdks/go/examples/oneOfEverything/value_handlers.go (hunk 16 above)
Comment (ctwarog-ross) [reply sdks/go/examples/oneOfEverything/value_handlers.go]: its not technically unscoped, the comment is a bit misleading, it just means any read/write scope, not particular ones like in slot 0/1

File: sdks/go/examples/oneOfEverything/value_handlers.go (hunk 10 above)
Comment (ctwarog-ross) [reply sdks/go/examples/oneOfEverything/value_handlers.go]: Thewre have been changes to ToProto and unit tests to ensure it always deep copies data. Cloning help methods removed from example.

File: sdks/go/examples/hello_world/main.go [hunk 26]
```diff
@@ -0,0 +1,183 @@
...
+	return s.value
+}
+
+func (s *helloWorldState) Set(value string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.value = value
+}
+
+func normalizeFqoid(fqoid string) string {
```
Comment (abrown-RV) [inline sdks/go/examples/hello_world/main.go]: remove here too

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 27]
```diff
@@ -0,0 +1,611 @@
...
+		var v any
+		var ok bool
+		switch fqoid {
+		case "product", "product/name", "product/vendor", "product/version":
+			v, ok = state.slotZeroProductValue(fqoid)
+		case "counter":
+			v, ok = counter.GetValue(), true
+		case "running":
+			v, ok = counter.RunningInt32(), true
+		case "sample_int_array":
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: should this array go live with the other arrays in slot 2?

File: sdks/go/examples/oneOfEverything/value_handlers.go (hunk 16 above)
Comment (abrown-RV) [reply sdks/go/examples/oneOfEverything/value_handlers.go]: That's still not to spec: every param must only have a single scope associtated with it, either directly on the param or inheirted from the parent/device default. The idea is that users have multiple scopes assigned to them to acheive broad access instead of params having multiple scopes. It does feel backwards at first, but thats the way it is. HOWEVER, here in L1 it does get muddy, and I fully expect devices will exist that don't really follow this. It just conflicts with the DeviceRequest/GetParam reporting the single scope on a param. So all that being said I don't know what to say for doing scopes here in the handlers, maybe it really is better to just do the device level default, check that at the top of the handler and call it a day, cause otherwise *every* if branch needs its own check. If we do that route, that the GetValue also needs the same scope checks and the DeviceRequest should report the same scope for each.

**Review by abrown-RV — CHANGES_REQUESTED** [review]
All these var types in slot 2 is going to be great for the Check Your Catena tool

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 28]
```diff
@@ -0,0 +1,611 @@
...
+			if idx, field, _, okPath := parseArrayElementPath(fqoid, "sample_int_array"); okPath && field == "" {
+				arr := state.sampleIntArrayValue()
+				if idx < len(arr) {
+					v, ok = arr[idx], true
+				}
+			}
+		}
+
+		return replyValue(fqoid, v, ok)
+	})
```
[the remark this answers is not in the document]
Comment (ctwarog-ross) [reply sdks/go/examples/oneOfEverything/value_handlers.go:36]: / was removed from spec. Disregard

File: sdks/go/examples/oneOfEverything/command_handler.go [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+			return catena.CommandReply(val)
+		case "reset":
+			counter.Reset()
+			logger.Info("Counter reset", "value", counter.GetValue())
+			srv.BroadcastUpdate(0, "counter", counter.GetValue(), counterScope)
+			val, _ := catena.ToValue(counter.GetValue())
+			return catena.CommandReply(val)
+		default:
+			logger.Warning("Command not found", "slot", slot, "command", commandFqoid)
+			return catena.CommandExceptionResult("Invalid Command", "Command not found: "+commandFqoid, nil)
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/command_handler.go]: should return a `StatusCodeNotFound` this is a protocol error not a command error.

File: sdks/go/examples/oneOfEverything/command_handler.go [hunk 30]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/rossvideo/catena/sdks/go/pkg/logger"
+)
+
+func registerCommandHandler(srv catena.Server, counter *CounterState, broadcastRunning func(), counterScope string) {
+	// Slot 0: switch per command FQOID and build command responses manually.
+	// This handler does not reduce allowed scopes beyond the base level of
+	// any command scope (mon, op, conf, or adm), however it broadcasts updates
+	// to the counter value only to specified counterScope.
+	srv.RegisterExecuteCommandHandler(0, func(slot uint16, commandFqoid string, payload any, ctx catena.HandlerContext) (catena.CommandResult, catena.StatusResult) {
+		logger.Info("ExecuteCommand", "slot", slot, "command", commandFqoid)
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/command_handler.go:14]:
```suggestion
		logger.Info("ExecuteCommand", "slot", slot, "command", commandFqoid)
		if !ctx.HasWriteScope(counterScope) {
			logger.Warning("Unauthorized command execution attempt", "slot", slot, "command", commandFqoid)
			return catena.ReplyError[catena.CommandResult](catena.StatusCodePermissionDenied, "Caller does not have required scope for this command")
		}
```
Should check scope

File: sdks/go/pkg/catena/value.go [hunk 31]
```diff
@@ -126,11 +127,11 @@ func ToProto(v any) (*protos.Value, StatusResult) {
 	case string:
 		return &protos.Value{Kind: &protos.Value_StringValue{StringValue: val}}, StatusResult{Code: StatusCodeOk}
 	case []int32:
-		return &protos.Value{Kind: &protos.Value_Int32ArrayValues{Int32ArrayValues: &protos.Int32List{Ints: val}}}, StatusResult{Code: StatusCodeOk}
+		return &protos.Value{Kind: &protos.Value_Int32ArrayValues{Int32ArrayValues: &protos.Int32List{Ints: slices.Clone(val)}}}, StatusResult{Code: StatusCodeOk}
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/value.go:138]: ToValue and BroadcastUpdate should get documentation indicating their cloning behaviour. Here's what copilot gave me:

```go
// ToValue converts v into a Value with its own copied backing data.
//
// For supported input types, the returned Value does not alias caller-owned
// slices, maps, or payload buffers, so callers may release locks and freely
// mutate or discard the original input after ToValue returns.
//
// Callers must still prevent concurrent mutation of v while this function is
// running.
func ToValue(v any) (Value, StatusResult) {
```

File: sdks/go/examples/hello_world/main.go [hunk 32]
```diff
@@ -0,0 +1,178 @@
...
+
+	logger.Info("=======================================================")
+	logger.Info("Hello World Example")
+	logger.Info("=======================================================")
+	logger.Info("REST transport listening on default port 8080")
+	logger.Info("gRPC transport listening on default port 6254")
+	logger.Info("Try REST:")
+	logger.Info("  GET  http://localhost:8080/st2138-api/v1/0/value/hello_world")
+	logger.Info("Try gRPC:")
+	logger.Info("  grpcurl -plaintext -d '{\"slot\":0,\"oid\":\"hello_world\"}' localhost:6254 st2138.CatenaService/ValueRequest")
```
[the remark this answers is not in the document]
Comment (abrown-RV) [reply sdks/go/examples/hello_world/main.go]: bugbot is right, what a deep cut

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 33]
```diff
@@ -0,0 +1,678 @@
...
+				state.mu.RUnlock()
+				return status
+			}
+		}
+		state.mu.RUnlock()
+
+		state.mu.Lock()
+		defer state.mu.Unlock()
+		for _, entry := range entries {
+			if status := slotTwoApplySet(entry.Fqoid, entry.Value, state); status.Code != catena.StatusCodeOk {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go]: this should not fail, validate should ensure that apply cannot fail

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 34]
```diff
@@ -0,0 +1,678 @@
...
+	// type. This slot only requires monitor-scope write access, demonstrating
+	// that each handler can enforce a different authorization policy.
+	srv.RegisterSetValueHandler(1, func(slot uint16, entries []catena.SetValueEntry, ctx catena.HandlerContext) catena.StatusResult {
+		logger.Info("SetValue", "slot", slot, "count", len(entries))
+		if !ctx.HasWriteScope(catena.ScopeMon) {
+			return catena.StatusWithCode(catena.StatusCodePermissionDenied, "monitor scope required")
+		}
+
+		for _, entry := range entries {
+			if entry.Value == nil {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go:160]: wonder if we can do something in the sdk that guarentees it will never be called with nil? so this boilerplate isn't needed everywhere? I guess since its an any, probably not...

**Review by abrown-RV — COMMENTED** [review]
some more bugbot things too

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 35]
```diff
@@ -199,11 +198,12 @@ func registerValueHandlers(srv catena.Server, counter *CounterState, state *Exam
...
-		defer state.mu.Unlock()
 		for _, entry := range entries {
-			if status := slotTwoApplySet(entry.Fqoid, entry.Value, state); status.Code != catena.StatusCodeOk {
-				return status
-			}
+			slotTwoApplySet(entry.Fqoid, entry.Value, state)
+		}
+		state.mu.Unlock()
+
+		for _, entry := range entries {
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go:212]: Shouldn't BroadcastUpdate happen under lock? So it can clone the unmodified value

File: sdks/go/examples/oneOfEverything/value_handlers.go [hunk 36]
```diff
@@ -199,11 +198,12 @@ func registerValueHandlers(srv catena.Server, counter *CounterState, state *Exam
 		state.mu.RUnlock()
 
 		state.mu.Lock()
-		defer state.mu.Unlock()
 		for _, entry := range entries {
-			if status := slotTwoApplySet(entry.Fqoid, entry.Value, state); status.Code != catena.StatusCodeOk {
-				return status
-			}
+			slotTwoApplySet(entry.Fqoid, entry.Value, state)
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/value_handlers.go:210]: Isn't there technically a window here between the RUnlock and the Lock where something could step in and change things?

File: sdks/go/hello_world
Comment (abrown-RV) [inline sdks/go/hello_world:1]: accidental binary

File: sdks/go/examples/oneOfEverything/value_handlers.go (hunk 36 above)
Comment (ctwarog-ross) [reply sdks/go/examples/oneOfEverything/value_handlers.go:210]: yes, a lock should be carried through the whole handler

**Review by abrown-RV — DISMISSED** [review]

**Comment by ctwarog-ross** [comment]
bugbot run

**Review by abrown-RV — APPROVED** [review]
